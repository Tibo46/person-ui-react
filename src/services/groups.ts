import * as t from 'io-ts';
import { get } from '../utils/http';

export const Group = t.strict({
  id: t.string,
  name: t.string,
});
export type Group = t.TypeOf<typeof Group>;

export const GroupResponse = t.array(Group);

export type GroupResponse = t.TypeOf<typeof GroupResponse>;
export async function getTypes(): Promise<Group[]> {
  const res = await get(GroupResponse, {
    url: `${window.ENV_DATA.personApiOrigin}/groups`,
  });

  return res;
}
