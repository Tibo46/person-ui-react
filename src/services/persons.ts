import * as t from 'io-ts';
import { get } from '../utils/http';
import { Group } from './groups';

export const Person = t.strict({
  id: t.string,
  name: t.string,
  group: Group,
});
export type Person = t.TypeOf<typeof Person>;

export const SearchResponse = t.array(Person);

export type SearchResponse = t.TypeOf<typeof SearchResponse>;

interface SearchPersonQuery {
  search: string;
}
export async function searchPersons({ query }: { query: SearchPersonQuery }): Promise<Person[]> {
  const searchTerms = query.search ? `?searchTerms=${query.search}` : '';

  const res = await get(SearchResponse, {
    url: `${window.ENV_DATA.personApiOrigin}/persons/search${searchTerms}`,
  });

  return res;
}

interface GetPersonQuery {
  id: string;
}
export async function getPerson({ query }: { query: GetPersonQuery }): Promise<Person> {
  const res = await get(Person, {
    url: `${window.ENV_DATA.personApiOrigin}/persons/${query.id}`,
  });

  return res;
}
