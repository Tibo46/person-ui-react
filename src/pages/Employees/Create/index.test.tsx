import React from 'react';
import { act } from 'react-dom/test-utils';
import fetchMock from 'fetch-mock';

import renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';
import CreateEmployee from '.';
import { GroupResponse } from '../../../services/groups';

let mockHistory: any;
jest.mock('react-router', () => ({ useHistory: () => mockHistory }));

const delay = (time = 0) => new Promise((res) => setTimeout(res, time));

window.ENV_DATA = {
  personApiOrigin: 'https://mock-api.com',
} as any;

describe('Create Employee page', () => {
  let allGroupsResult: GroupResponse;

  beforeEach(() => {
    mockHistory = { push: jest.fn() };
    allGroupsResult = [
      {
        id: 1,
        name: 'Chemist',
      },
      {
        id: 2,
        name: 'Commercial',
      },
    ];
    fetchMock.get(`${window.ENV_DATA.personApiOrigin}/groups`, allGroupsResult);
    fetchMock.post(`${window.ENV_DATA.personApiOrigin}/persons`, {});
  });

  afterEach(() => {
    fetchMock.reset();
  });

  it('renders the page', async () => {
    const component = renderer.create(<CreateEmployee />);

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('should render a loader while fetching data and then display the data', async () => {
    let component: renderer.ReactTestRenderer;
    renderer.act(() => {
      component = renderer.create(<CreateEmployee />);
    });

    expect(component!.toJSON()).toMatchSnapshot();

    await renderer.act(async () => {
      await delay();
    });

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('should display a loader and then an error message if the api is down', async () => {
    fetchMock.get(`${window.ENV_DATA.personApiOrigin}/groups`, 500, { overwriteRoutes: true });

    let component: renderer.ReactTestRenderer;
    renderer.act(() => {
      component = renderer.create(<CreateEmployee />);
    });

    expect(component!.toJSON()).toMatchSnapshot();

    await renderer.act(async () => {
      await delay();
    });

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('should not post to the API when submitting the form if name is empty', async () => {
    let component: ReactWrapper;
    await act(async () => {
      component = mount(<CreateEmployee />);
    });

    expect(fetchMock.calls()).toHaveLength(1);

    await act(async () => {
      await delay();
    });
    component!.update();

    await act(async () => {
      component.find('form').simulate('submit');
    });

    expect(fetchMock.calls()).toHaveLength(1);
  });

  it('should post to the API when submitting the form', async () => {
    let component: ReactWrapper;
    await act(async () => {
      component = mount(<CreateEmployee />);
    });

    expect(fetchMock.calls()).toHaveLength(1);

    await act(async () => {
      await delay();
    });
    component!.update();

    await act(async () => {
      component
        .find('input[name="employee-name"]')
        .simulate('change', { target: { value: 'Gustavo Fring' } });
      component
        .find('input[name="employee-photo"]')
        .simulate('change', { target: { value: 'https://mock-img/grumpycat.jpg' } });
    });
    component!.update();
    await act(async () => {
      component.find('form').simulate('submit');
    });

    expect(fetchMock.calls()).toHaveLength(2);
    expect(fetchMock.calls()[1][0]).toBe('https://mock-api.com/persons');
    expect(fetchMock.calls()[1][1]?.body).toBe(
      JSON.stringify({ name: 'Gustavo Fring', groupId: 1, photo: 'https://mock-img/grumpycat.jpg' })
    );
  });

  it('should redirect the user af', async () => {
    let component: ReactWrapper;
    await act(async () => {
      component = mount(<CreateEmployee />);
    });

    await act(async () => {
      await delay();
    });
    component!.update();

    expect(mockHistory.push).not.toHaveBeenCalled();

    await act(async () => {
      component
        .find('input[name="employee-name"]')
        .simulate('change', { target: { value: 'Gustavo Fring' } });
    });
    component!.update();
    await act(async () => {
      component.find('form').simulate('submit');
    });

    expect(mockHistory.push).toHaveBeenCalled();
  });
});
