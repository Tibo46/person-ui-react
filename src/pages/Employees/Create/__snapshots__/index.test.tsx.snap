// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Employee page renders the page 1`] = `
Array [
  <h1
    className="MuiTypography-root MuiTypography-h1"
  >
    Add a new employee
  </h1>,
  <div
    className="makeStyles-loader"
  />,
]
`;

exports[`Create Employee page should display a loader and then an error message if the api is down 1`] = `
Array [
  <h1
    className="MuiTypography-root MuiTypography-h1"
  >
    Add a new employee
  </h1>,
  <div
    className="makeStyles-loader"
  />,
]
`;

exports[`Create Employee page should display a loader and then an error message if the api is down 2`] = `
Array [
  <h1
    className="MuiTypography-root MuiTypography-h1"
  >
    Add a new employee
  </h1>,
  <p
    className="MuiTypography-root no-data MuiTypography-body1"
  >
    An error occured and we could not get the Groups. Please try again by refreshing the page.
  </p>,
]
`;

exports[`Create Employee page should render a loader while fetching data and then display the data 1`] = `
Array [
  <h1
    className="MuiTypography-root MuiTypography-h1"
  >
    Add a new employee
  </h1>,
  <div
    className="makeStyles-loader"
  />,
]
`;

exports[`Create Employee page should render a loader while fetching data and then display the data 2`] = `
Array [
  <h1
    className="MuiTypography-root MuiTypography-h1"
  >
    Add a new employee
  </h1>,
  <form
    autoComplete="off"
    className="container"
    onSubmit={[Function]}
  >
    <mui-grid
      container={true}
      direction="column"
      spacing={2}
    >
      <mui-grid
        container={true}
        item={true}
        spacing={2}
      >
        <mui-grid
          item={true}
          sm={6}
        >
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-required MuiInputLabel-required"
              data-shrink={false}
            >
              Employee name
              <span
                className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="off"
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                name="employee-name"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
                value=""
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled"
                >
                  <span>
                    Employee name
                     *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </mui-grid>
        <mui-grid
          item={true}
          sm={6}
        >
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled MuiFormLabel-required MuiInputLabel-required"
              data-shrink={true}
            >
              Employee group
              <span
                className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              onClick={[Function]}
            >
              <div
                aria-haspopup="listbox"
                aria-labelledby="mui-component-select-employee-group"
                className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
                id="mui-component-select-employee-group"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                role="button"
                tabIndex={0}
              >
                Chemist
              </div>
              <input
                autoFocus={false}
                name="employee-group"
                onAnimationStart={[Function]}
                type="hidden"
                value={1}
              />
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled PrivateNotchedOutline-legendNotched"
                >
                  <span>
                    Employee group
                     *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </mui-grid>
      </mui-grid>
      <mui-grid
        item={true}
        sm={12}
      >
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink={false}
          >
            Employee photo url
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoComplete="off"
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              name="employee-photo"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="https://myimagedomain.com/grumpycat.jpg"
              required={false}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled"
              >
                <span>
                  Employee photo url
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </mui-grid>
      <mui-grid
        item={true}
      >
        <mui-button
          color="primary"
          disableRipple={true}
          size="large"
          type="submit"
          variant="contained"
        >
          Create
        </mui-button>
      </mui-grid>
    </mui-grid>
  </form>,
]
`;
